File: src/condition/Condition.php
<?php


namespace Ething\Condition;

use \Ething\RuleItem;
use \Ething\InvalidRuleException;
use \Ething\Event\Signal;

abstract class Condition extends RuleItem {
	
	
	final public function evaluate(Signal $signal){
		if($this->isValid()){
			
			// reset error
			$this->setError(false);
			
			try {
				return boolval($this->call($signal));
			}
			catch(InvalidRuleException $e){
				// this exception is fired when an action is no more valid.
				$this->setInvalid();
				$this->setError($e->getMessage());
			}
			catch(\Exception $e){
				$this->setError($e->getMessage());
			}
			
		}
		
		return false;
	}
	
	abstract protected function call(\Ething\Event\Signal $signal);
	
	
}


	

File: ething/rule/Condition.py
# coding: utf-8

from .RuleItem import RuleItem
from . import InvalidRuleException
from ething.meta import MetaCondition
from future.utils import with_metaclass


class Condition(with_metaclass(MetaCondition,RuleItem)):
    
    
    def evaluate(self, signal):
        if self.valid:
            
            # reset error
            self.error = False
            
            try:
                return bool(self.call(signal))
            except InvalidRuleException as e:
                # this exception is fired when this condition is no more valid.
                self.valid = False
                self.error = e
                self.ething.log.exception("Fatal error in condition %s" % self)
            
            except Exception as e:
                self.error = e
                self.ething.log.exception("Error in condition %s" % self)
        
        return False
    
    
    def call(self, signal):
        raise NotImplementedError()
